### PR TITLE
Store git token as a secret in the cluster

### PR DESCRIPTION
**Pull request by josecordaz**
Fixes: #498 

**What changed?**
Initial manifests that api-service will need to set proper authorization permissions. There are still a few rules missing that will be added later when we have the api-service running in the cluster. 
Skip git token prompt if there is a value in the env GITHUB_TOKEN.
Added secret deletion to the restart-wego.sh script that is used in acceptance tests.

**Why?**
When the api-service runs in the cluster we would need to git token to be consumed via secret.

**How did you test it?**
Acceptance tests passing in the CI.

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: pkg/kube/kube.go [hunk 1]
```diff
@@ -226,3 +231,44 @@ func (k *KubeClient) LabelExistsInCluster(ctx context.Context, label string) err
 	}
 	return nil
 }
+
+func (k *KubeClient) CreateSecret(ctx context.Context, name, key, value string, namespace string) error {
+	cmd := []string{
+		"create",
+		"secret",
+		"generic",
+		name,
```
Comment (J-Thompson12) [inline pkg/kube/kube.go:240]: Does this mean that only one git token can be stored in the cluster?

File: pkg/services/gitops/install.go [hunk 2]
```diff
@@ -32,9 +33,22 @@ func (g *Gitops) Install(params InstallParams) ([]byte, error) {
...
 	} else {
+		ctx := context.Background()
+		if err := g.kube.CreateSecret(ctx, kube.GitTokenSecretName, kube.GitTokenKeyName, params.GitToken, params.Namespace); err != nil {
+			return nil, errors.Wrap(err, "could not create git token secret")
+		}
 		if out, err := g.kube.Apply(manifests.AppCRD, params.Namespace); err != nil {
 			return []byte{}, errors.Wrapf(err, "failed to apply App spec CR: %s", string(out))
 		}
+		if out, err := g.kube.Apply(manifests.ServiceAccountApiService, params.Namespace); err != nil {
+			return []byte{}, errors.Wrapf(err, "failed to apply service account manifest for api-service: %s", string(out))
```
Comment (luizbafilho) [inline pkg/services/gitops/install.go:44]: after some discussion with Kevin we've decided to move to `fmt.Errorf`. You don't need to change all the codebase, but as we work through the packages we should take the chance to change the calls there.

You can see more about it: https://blog.golang.org/go1.13-errors

**Review by jpellizzari — COMMENTED** [review]
Nice work! Let me know if you want to pair on the `KubeHTTP` implementations.

**Review by luizbafilho — CHANGES_REQUESTED** [review]

File: pkg/kube/kube.go [hunk 3]
```diff
@@ -59,6 +62,8 @@ type Kube interface {
 	LabelExistsInCluster(ctx context.Context, label string) error
 	GetApplication(ctx context.Context, name types.NamespacedName) (*wego.Application, error)
 	GetResource(ctx context.Context, name types.NamespacedName, resource Resource) error
+	CreateSecret(ctx context.Context, name, key, value string, namespace string) error
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply pkg/kube/kube.go:65]: > Just as a convention define the type of all the params
> 
> `CreateSecret(ctx context.Context, name string, key string, value string, namespace string) error`

We need a linting rule to enforce this if we decide to enforce it as group

[the remark this answers is not in the document]
Comment (luizbafilho) [reply pkg/kube/kube.go:65]: agree

File: manifests/api-service/role.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: secret-reader
+  namespace: wego-system
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply manifests/api-service/role.yaml:5]: @josecordaz Might be easier to create a role via the `go-client` than using templating? Just an idea:
https://pkg.go.dev/k8s.io/api/rbac/v1beta1#Role

[the remark this answers is not in the document]
Comment (jpellizzari) [reply manifests/api-service/role.yaml:5]: If we need to store them in the user's repo, the k8s `yaml` library can help with converting the struct to bytes.

[the remark this answers is not in the document]
Comment (luizbafilho) [reply manifests/api-service/role.yaml:5]: I don't think we need that versioned just like we don't version the manifests from flux install, but we definitely need to output them when the user select the dry run in the install

File: pkg/kube/kube.go [hunk 5]
```diff
@@ -226,3 +231,44 @@ func (k *KubeClient) LabelExistsInCluster(ctx context.Context, label string) err
...
+
+	_, err := k.runKubectlCmd(cmd)
+	if err != nil {
+		return fmt.Errorf("could not run kubectl command: %s", err)
+	}
+
+	return nil
+}
+
+func (k *KubeClient) GetSecret(ctx context.Context, name, key string, namespace string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (josecordaz) [reply pkg/kube/kube.go:254]: @luizbafilho not really sure what do you mean, would you mind to clarify please?

File: pkg/kube/kube.go (hunk 1 above)
Comment (josecordaz) [reply pkg/kube/kube.go:240]: The logic currently needs only one git token. This `name` variable could be any valid string. We could use it to create as many secrets as we need.

File: cmd/wego/gitops/cmd.go [hunk 6]
```diff
@@ -75,11 +78,25 @@ func installRunCmd(cmd *cobra.Command, args []string) error {
 	fluxClient := flux.New(cliRunner)
 	kubeClient := kube.New(cliRunner)
 
+	var gitToken string
+	gitToken = os.Getenv("GITHUB_TOKEN")
+	if gitToken == "" {
+		fmt.Print("Github token: ")
```
[the remark this answers is not in the document]
Comment (josecordaz) [reply cmd/wego/gitops/cmd.go:84]: Agree, what about "Github token will be used to create deployment keys and pull requests: " ? @jpellizzari

[the remark this answers is not in the document]
Comment (jpellizzari) [reply cmd/wego/gitops/cmd.go:84]: Sounds good!

File: manifests/api-service/role.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (josecordaz) [reply manifests/api-service/role.yaml]: I removed the namespace fields from all the manifests files. As we are already passing in the namespace to the apply command the manifest will be created in the namespace specified by it.

File: pkg/kube/kube.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (josecordaz) [reply pkg/kube/kube.go]: nvm I just figured it out

**Comment by josecordaz** [comment]
@jpellizzari @luizbafilho changes were addressed, would you mind to take a second look

**Review by luizbafilho — APPROVED** [review]

File: cmd/wego/gitops/cmd.go [hunk 7]
```diff
@@ -75,11 +78,25 @@ func installRunCmd(cmd *cobra.Command, args []string) error {
...
+	gitProviderToken = os.Getenv("GITHUB_TOKEN")
+	if gitProviderToken == "" {
+		fmt.Print("Github token will be used to create deployment keys and pull requests: ")
+		input, err := term.ReadPassword(int(os.Stdin.Fd()))
+		if err != nil {
+			return errors.Wrap(err, "failed reading ssh key password")
+		}
+		gitProviderToken = string(input)
+	} else {
+		fmt.Println("Using env GITHUB_TOKEN")
```
Comment (jpellizzari) [inline cmd/wego/gitops/cmd.go:91]: Sorry I should have caught this before. We need to be explicit about what we are doing. I suggest:

`Copying GITHUB_TOKEN from env to cluster secret for create deploy keys and pull requests`.

A little verbose maybe, but we need to be explicit about this.

File: cmd/wego/gitops/cmd.go [hunk 8]
```diff
@@ -75,11 +78,25 @@ func installRunCmd(cmd *cobra.Command, args []string) error {
 	fluxClient := flux.New(cliRunner)
 	kubeClient := kube.New(cliRunner)
 
+	var gitProviderToken string
+	gitProviderToken = os.Getenv("GITHUB_TOKEN")
+	if gitProviderToken == "" {
+		fmt.Print("Github token will be used to create deployment keys and pull requests: ")
```
Comment (jpellizzari) [inline cmd/wego/gitops/cmd.go:84]:
```suggestion
		fmt.Print("Github token will be used to create deploy keys and pull requests: ")
```
"Deploy Keys" is a specific thing in GitHub.

Comment (jpellizzari) [reply cmd/wego/gitops/cmd.go:84]: Also should this be `Println`?

**Review by jpellizzari — COMMENTED** [review]
The PR itself looks good.

We are still evaluating this auth approach against some others. Please hold off on merging until we discuss 1-1.

**Comment by jpellizzari** [comment]
Closing this in favor of the approach laid out here: https://github.com/weaveworks/weave-gitops/pull/513

**Pull request closed without merging**